File: swift/Sources/Fory/Decimal.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

private let decimalSmallPositiveMax: UInt64 = 0x3fff_ffff_ffff_ffff
private let decimalSmallNegativeAbsMax: UInt64 = 0x4000_0000_0000_0000
private let decimalLengthMask: UInt8 = 0x0f
private let decimalNegativeMask: UInt8 = 0x10
private let decimalCompactMask: UInt8 = 0x20
private let decimalHeaderSize = 4
private let decimalMaxMantissaWords = 8
private let decimalMaxMagnitudeBytes = decimalMaxMantissaWords * 2

private struct FoundationDecimalWireState {
    let scale: Int32
    let signum: Int8
    let magnitude: [UInt8]
}

@inline(__always)
private func normalizeDecimalMagnitude(_ magnitude: [UInt8]) -> [UInt8] {
    var normalized = magnitude
    while let last = normalized.last, last == 0 {
        normalized.removeLast()
    }
    return normalized
}

@inline(__always)
private func decimalMagnitudeBytes(from value: UInt64) -> [UInt8] {
    guard value != 0 else {
        return []
    }
    var remaining = value
    var bytes: [UInt8] = []
    bytes.reserveCapacity(8)
    while remaining != 0 {
        bytes.append(UInt8(truncatingIfNeeded: remaining))
        remaining >>= 8
    }
    return bytes
}

@inline(__always)
private func decimalUInt64(from magnitude: [UInt8]) -> UInt64? {
    guard magnitude.count <= 8 else {
        return nil
    }
    var result: UInt64 = 0
    for (index, byte) in magnitude.enumerated() {
        result |= UInt64(byte) << (index * 8)
    }
    return result
}

private func divideDecimalMagnitudeBy10(_ magnitude: inout [UInt8]) -> UInt8 {
    guard !magnitude.isEmpty else {
        return 0
    }
    var remainder = 0
    for index in stride(from: magnitude.count - 1, through: 0, by: -1) {
        let value = remainder * 256 + Int(magnitude[index])
        magnitude[index] = UInt8(value / 10)
        remainder = value % 10
    }
    magnitude = normalizeDecimalMagnitude(magnitude)
    return UInt8(remainder)
}

private func decimalMagnitudeString(signum: Int8, magnitude: [UInt8]) -> String {
    let normalized = normalizeDecimalMagnitude(magnitude)
    guard !normalized.isEmpty else {
        return "0"
    }

    var remainderMagnitude = normalized
    var digits: [Character] = []
    while !remainderMagnitude.isEmpty {
        let digit = divideDecimalMagnitudeBy10(&remainderMagnitude)
        digits.append(Character(String(UnicodeScalar(48 + Int(digit))!)))
    }
    let decimalDigits = String(digits.reversed())
    return signum < 0 ? "-\(decimalDigits)" : decimalDigits
}

@inline(__always)
private func encodeDecimalZigZag64(_ value: Int64) -> UInt64 {
    UInt64(bitPattern: (value << 1) ^ (value >> 63))
}

@inline(__always)
private func decodeDecimalZigZag64(_ value: UInt64) -> Int64 {
    let shifted = Int64(bitPattern: value >> 1)
    let mask = -Int64(value & 1)
    return shifted ^ mask
}

private func decimalMantissaWords(from magnitude: [UInt8]) throws -> (words: [UInt16], length: Int) {
    let normalized = normalizeDecimalMagnitude(magnitude)
    guard normalized.count <= decimalMaxMagnitudeBytes else {
        throw ForyError.invalidData(
            "decimal magnitude with \(normalized.count) bytes exceeds Foundation.Decimal precision"
        )
    }
    var words = Array(repeating: UInt16(0), count: decimalMaxMantissaWords)
    for (index, byte) in normalized.enumerated() {
        let wordIndex = index / 2
        let shift = (index % 2) * 8
        words[wordIndex] |= UInt16(byte) << shift
    }
    var length = words.count
    while length > 0, words[length - 1] == 0 {
        length -= 1
    }
    return (words, length)
}

private func foundationDecimalExponent(forScale scale: Int32) throws -> Int8 {
    let exponent = 0 - Int64(scale)
    guard exponent >= Int64(Int8.min), exponent <= Int64(Int8.max) else {
        throw ForyError.invalidData(
            "decimal scale \(scale) is out of Foundation.Decimal exponent range"
        )
    }
    return Int8(exponent)
}

private func foundationDecimalWireState(_ value: Decimal) -> FoundationDecimalWireState {
    var compact = value
    NSDecimalCompact(&compact)
    return withUnsafeBytes(of: &compact) { raw in
        let exponent = Int8(bitPattern: raw[0])
        let flags = raw[1]
        let length = min(Int(flags & decimalLengthMask), decimalMaxMantissaWords)
        let isNegative = (flags & decimalNegativeMask) != 0

        var magnitude: [UInt8] = []
        magnitude.reserveCapacity(length * 2)
        for index in 0..<length {
            let offset = decimalHeaderSize + index * 2
            magnitude.append(raw[offset])
            magnitude.append(raw[offset + 1])
        }
        let normalized = normalizeDecimalMagnitude(magnitude)
        return FoundationDecimalWireState(
            scale: Int32(0) - Int32(exponent),
            signum: normalized.isEmpty ? 0 : (isNegative ? -1 : 1),
            magnitude: normalized
        )
    }
}

private func buildFoundationDecimal(
    signum: Int8,
    magnitude: [UInt8],
    scale: Int32
) throws -> Decimal {
    let normalized = normalizeDecimalMagnitude(magnitude)
    let exponent = try foundationDecimalExponent(forScale: scale)
    let mantissa = try decimalMantissaWords(from: normalized)

    var value = Decimal.zero
    withUnsafeMutableBytes(of: &value) { raw in
        raw.initializeMemory(as: UInt8.self, repeating: 0)
        raw[0] = UInt8(bitPattern: exponent)
        var flags = UInt8(truncatingIfNeeded: mantissa.length)
        if signum < 0 && !normalized.isEmpty {
            flags |= decimalNegativeMask
        }
        if mantissa.length > 0 {
            flags |= decimalCompactMask
        }
        raw[1] = flags
        for index in 0..<decimalMaxMantissaWords {
            let offset = decimalHeaderSize + index * 2
            let word = mantissa.words[index]
            raw[offset] = UInt8(truncatingIfNeeded: word)
            raw[offset + 1] = UInt8(truncatingIfNeeded: word >> 8)
        }
    }
    NSDecimalCompact(&value)
    return value
}

private func smallUnscaledValueForWire(_ state: FoundationDecimalWireState) -> Int64? {
    guard let magnitude = decimalUInt64(from: state.magnitude) else {
        return nil
    }
    if state.signum >= 0 {
        guard magnitude <= decimalSmallPositiveMax else {
            return nil
        }
        return Int64(magnitude)
    }
    guard magnitude <= decimalSmallNegativeAbsMax else {
        return nil
    }
    return -Int64(magnitude)
}

extension Decimal {
    internal var foryScale: Int32 {
        foundationDecimalWireState(self).scale
    }

    internal var foryUnscaledString: String {
        let state = foundationDecimalWireState(self)
        return decimalMagnitudeString(signum: state.signum, magnitude: state.magnitude)
    }
}

extension Decimal: Serializer {
    public static func defaultValue(_ context: ReadContext) throws -> Decimal {
        .zero
    }

    public static var staticTypeId: TypeId {
        .decimal
    }

    public static func writeData(_ value: Self, _ context: WriteContext) throws {
        let state = foundationDecimalWireState(value)
        context.buffer.writeVarInt32(state.scale)
        if let small = smallUnscaledValueForWire(state) {
            let header = encodeDecimalZigZag64(small) << 1
            context.buffer.writeVarUInt64(header)
            return
        }

        guard !state.magnitude.isEmpty else {
            throw ForyError.invalidData("zero must use the small decimal encoding")
        }
        let sign: UInt64 = state.signum < 0 ? 1 : 0
        let meta = (UInt64(state.magnitude.count) << 1) | sign
        let header = (meta << 1) | 1
        context.buffer.writeVarUInt64(header)
        context.buffer.writeBytes(state.magnitude)
    }

    public static func readData(_ context: ReadContext) throws -> Decimal {
        let scale = try context.buffer.readVarInt32()
        let header = try context.buffer.readVarUInt64()
        if (header & 1) == 0 {
            let unscaled = decodeDecimalZigZag64(header >> 1)
            let signum: Int8 = unscaled == 0 ? 0 : (unscaled < 0 ? -1 : 1)
            return try buildFoundationDecimal(
                signum: signum,
                magnitude: decimalMagnitudeBytes(from: unscaled.magnitude),
                scale: scale
            )
        }

        let meta = header >> 1
        let signum: Int8 = (meta & 1) == 0 ? 1 : -1
        let rawLength = meta >> 1
        guard rawLength > 0 else {
            throw ForyError.invalidData("invalid decimal magnitude length \(rawLength)")
        }
        // Foundation.Decimal has eight 16-bit mantissa words. Check the unsigned
        // wire length before native conversion or copying an attacker-sized body.
        guard rawLength <= UInt64(decimalMaxMagnitudeBytes) else {
            throw ForyError.invalidData(
                "decimal magnitude with \(rawLength) bytes exceeds Foundation.Decimal precision"
            )
        }
        let length = Int(rawLength)
        let magnitudeBytes = try context.buffer.readBytes(count: length)
        guard magnitudeBytes[length - 1] != 0 else {
            throw ForyError.invalidData("non-canonical decimal magnitude bytes: trailing zero byte")
        }
        let normalized = normalizeDecimalMagnitude(magnitudeBytes)
        guard !normalized.isEmpty else {
            throw ForyError.invalidData("big decimal encoding must not represent zero")
        }
        return try buildFoundationDecimal(signum: signum, magnitude: normalized, scale: scale)
    }
}

File: go/fory/decimal.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"fmt"
	"math/big"
	"reflect"
)

// Decimal is the Go carrier for Fory DECIMAL values.
// It preserves the exact value unscaled * 10^-scale.
type Decimal struct {
	Unscaled big.Int
	Scale    int32
}

func NewDecimal(unscaled *big.Int, scale int32) Decimal {
	var copied big.Int
	if unscaled != nil {
		copied.Set(unscaled)
	}
	return Decimal{Unscaled: copied, Scale: scale}
}

func (d Decimal) Equal(other Decimal) bool {
	return d.Scale == other.Scale && d.Unscaled.Cmp(&other.Unscaled) == 0
}

func (d Decimal) String() string {
	return fmt.Sprintf("%se%d", d.Unscaled.String(), -d.Scale)
}

var (
	decimalReflectType = reflect.TypeFor[Decimal]()
	decimalLongMin     = big.NewInt(MinInt64)
	decimalLongMax     = big.NewInt(MaxInt64)
)

const maxDecimalMagnitudeBytes = 10_000
const maxDecimalScale int32 = 10_000

type decimalSerializer struct{}

func (s decimalSerializer) Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value) {
	decimal := value.Interface().(Decimal)
	if !validateDecimal(ctx.Err(), decimal.Scale, &decimal.Unscaled) {
		return
	}
	if refMode != RefModeNone {
		ctx.buffer.WriteInt8(NotNullValueFlag)
	}
	if writeType {
		ctx.buffer.WriteUint8(uint8(DECIMAL))
	}
	writeValidDecimalParts(ctx.buffer, decimal.Scale, &decimal.Unscaled)
}

func (s decimalSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
	decimal := value.Interface().(Decimal)
	writeDecimalParts(ctx, decimal.Scale, &decimal.Unscaled)
}

func (s decimalSerializer) Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value) {
	err := ctx.Err()
	if refMode != RefModeNone {
		if ctx.buffer.ReadInt8(err) == NullFlag {
			value.Set(reflect.Zero(value.Type()))
			return
		}
	}
	if readType && !ctx.readExpectedTypeID(DECIMAL) {
		return
	}
	if ctx.HasError() {
		return
	}
	s.ReadData(ctx, value)
}

func (s decimalSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
	scale, unscaled := readDecimalParts(ctx)
	if ctx.HasError() {
		return
	}
	value.Set(reflect.ValueOf(NewDecimal(unscaled, scale)))
}

func (s decimalSerializer) ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value) {
	s.Read(ctx, refMode, false, false, value)
}

func writeDecimalParts(ctx *WriteContext, scale int32, unscaled *big.Int) {
	if !validateDecimal(ctx.Err(), scale, unscaled) {
		return
	}
	writeValidDecimalParts(ctx.buffer, scale, unscaled)
}

// Root writers call this before their protocol header so a rejected direct
// Decimal cannot append a partial root to a caller-owned buffer.
func validateRootDecimal(ctxErr *Error, value any) bool {
	switch decimal := value.(type) {
	case Decimal:
		return validateDecimal(ctxErr, decimal.Scale, &decimal.Unscaled)
	case *Decimal:
		return decimal == nil || validateDecimal(ctxErr, decimal.Scale, &decimal.Unscaled)
	default:
		return true
	}
}

func validateDecimal(ctxErr *Error, scale int32, unscaled *big.Int) bool {
	if scale < -maxDecimalScale || scale > maxDecimalScale {
		ctxErr.SetError(SerializationErrorf(
			"decimal scale %d exceeds supported range [%d, %d]",
			scale, -maxDecimalScale, maxDecimalScale))
		return false
	}
	if unscaled != nil && unscaled.BitLen() > maxDecimalMagnitudeBytes*8 {
		ctxErr.SetError(SerializationErrorf(
			"decimal magnitude exceeds %d bytes", maxDecimalMagnitudeBytes))
		return false
	}
	return true
}

func writeValidDecimalParts(buffer *ByteBuffer, scale int32, unscaled *big.Int) {
	if unscaled == nil {
		unscaled = new(big.Int)
	}
	small := canUseSmallDecimalEncoding(unscaled)
	buffer.WriteVarint32(scale)
	if small {
		smallValue := unscaled.Int64()
		header := encodeDecimalZigZag64(smallValue) << 1
		buffer.WriteVarUint64(header)
		return
	}

	abs := new(big.Int).Abs(unscaled)
	magnitudeBytes := abs.Bytes()
	if len(magnitudeBytes) == 0 {
		panic("decimal zero must use the small encoding")
	}
	reverseBytes(magnitudeBytes)
	meta := (uint64(len(magnitudeBytes)) << 1) | uint64(signBit(unscaled.Sign()))
	buffer.WriteVarUint64((meta << 1) | 1)
	buffer.WriteBinary(magnitudeBytes)
}

func readDecimalParts(ctx *ReadContext) (int32, *big.Int) {
	err := ctx.Err()
	scale := ctx.buffer.ReadVarint32(err)
	if ctx.HasError() {
		return 0, nil
	}
	if scale < -maxDecimalScale || scale > maxDecimalScale {
		ctx.SetError(DeserializationErrorf(
			"decimal scale %d exceeds supported range [%d, %d]",
			scale, -maxDecimalScale, maxDecimalScale))
		return 0, nil
	}
	header := ctx.buffer.ReadVarUint64(err)
	if ctx.HasError() {
		return 0, nil
	}
	if (header & 1) == 0 {
		return scale, big.NewInt(decodeDecimalZigZag64(header >> 1))
	}

	meta := header >> 1
	length := meta >> 1
	if length == 0 {
		ctx.SetError(DeserializationErrorf("invalid decimal magnitude length %d", length))
		return 0, nil
	}
	if length > uint64(MaxInt) {
		ctx.SetError(DeserializationErrorf("invalid decimal magnitude length %d", length))
		return 0, nil
	}
	if length > maxDecimalMagnitudeBytes {
		ctx.SetError(DeserializationErrorf(
			"decimal magnitude length %d exceeds limit %d",
			length, maxDecimalMagnitudeBytes))
		return 0, nil
	}
	magnitudeBytes := ctx.buffer.ReadBytes(int(length), err)
	if ctx.HasError() {
		return 0, nil
	}
	if magnitudeBytes[len(magnitudeBytes)-1] == 0 {
		ctx.SetError(DeserializationError("non-canonical decimal magnitude bytes: trailing zero byte"))
		return 0, nil
	}
	bigEndian := append([]byte(nil), magnitudeBytes...)
	reverseBytes(bigEndian)
	magnitude := new(big.Int).SetBytes(bigEndian)
	if magnitude.Sign() == 0 {
		ctx.SetError(DeserializationError("big decimal encoding must not represent zero"))
		return 0, nil
	}
	if (meta & 1) != 0 {
		magnitude.Neg(magnitude)
	}
	return scale, magnitude
}

func canUseSmallDecimalEncoding(value *big.Int) bool {
	if value == nil {
		return true
	}
	if value.Cmp(decimalLongMin) < 0 || value.Cmp(decimalLongMax) > 0 {
		return false
	}
	return (encodeDecimalZigZag64(value.Int64()) & (1 << 63)) == 0
}

func encodeDecimalZigZag64(value int64) uint64 {
	return uint64((value << 1) ^ (value >> 63))
}

func decodeDecimalZigZag64(value uint64) int64 {
	return int64((value >> 1) ^ uint64(-(int64(value & 1))))
}

func signBit(sign int) int {
	if sign < 0 {
		return 1
	}
	return 0
}

func reverseBytes(values []byte) {
	for i, j := 0, len(values)-1; i < j; i, j = i+1, j-1 {
		values[i], values[j] = values[j], values[i]
	}
}
